python/ray/metrics_agent.py

```diff
@@ -0,0 +1,180 @@
+from opencensus.stats import measure as measure_module
+from opencensus.stats import stats as stats_module
+from opencensus.tags import tag_key as tag_key_module
+from opencensus.tags import tag_map as tag_map_module
+from opencensus.tags import tag_value as tag_value_module
+from opencensus.stats import view
+
+from ray import prometheus_exporter
+
+logger = logging.getLogger(__name__)
+
+
+# Since metrics agent only collects the point data from cpp processes,
+# we don't need any metrics types other than gauge.
+class Gauge(view.View):
+    def __init__(self, name, description, unit,
+                 tags: List[tag_key_module.TagKey]):
+        self._measure = measure_module.MeasureInt(name, description, unit)
+        self._view = view.View(name, description, tags, self.measure,
+                               aggregation.LastValueAggregation())
+
+    @property
+    def measure(self):
+        return self._measure
+
+    @property
+    def view(self):
+        return self._view
+
+    @property
+    def name(self):
+        return self.measure.name
+
+    @property
+    def description(self):
+        return self.measure.description
+
+    @property
+    def units(self):
+        return self.measure.unit
+
+    @property
+    def tags(self):
+        return self.view.columns
+
+    def __dict__(self):
+        return {
+            "name": self.measure.name,
+            "description": self.measure.description,
+            "units": self.measure.unit,
+            "tags": self.view.columns,
+        }
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        return str(self.__dict__())
+
+
+class MetricsAgent:
+    def __init__(self, metrics_export_port):
+        if not metrics_export_port:
+            metrics_export_port = 8888
+
+        # OpenCensus classes.
+        self.stats = stats_module.stats
+        self.view_manager = self.stats.view_manager
+        self.stats_recorder = self.stats.stats_recorder
+        # Port where we will expose metrics.
+        self.metrics_export_port = metrics_export_port
+        # Measure to record metrics.
+        self._measure_map = self.stats_recorder.new_measurement_map()
+        # metric name(str) -> view (view.View)
+        # It is updated dynamically as new metrics
+        # are delivered from cpp processes.
+        self._registry = defaultdict(lambda: None)
+        # Lock required because gRPC server uses multi-threads.
+        self._lock = threading.Lock()
+        # Whether or not there are metrics that are missing description and
+        # units information. This is used to dynamically update registry.
+        self._missing_information = False
+
+        # Configure exporter. (We currently only support prometheus).
+        self.view_manager.register_exporter(
+            prometheus_exporter.new_stats_exporter(
+                prometheus_exporter.Options(
+                    namespace="ray", port=metrics_export_port)))
```

Comment:
Could a user monkey-patch a different exporter in by just calling `self.register_exporter` on a custom exporter?

Reply:
@thomasdesr this is probably relevant to you :)

Reply:
Yeah. If users build their own sources, they can just change this part, and then it will work.